tests/cobs/src/main.c

```diff
@@ -28,33 +32,78 @@ static void roundtrip_test_runner(const void *input, const size_t length)
 	zassert_mem_equal(decoded_buffer, input, length);
 	zassert_equal(decoded_buffer[length], 0xAB);
 
+	uint8_t *const encoded_buffer2 = malloc(encoded_buffer_length);
+	uint8_t *const decoded_buffer2 = malloc(length + 1);
+	memset(encoded_buffer2, 0xAB, encoded_buffer_length);
+	decoded_buffer2[length] = 0xAB;
+
+	const size_t encoded_length2 =
+		cobs_encode_stream_simple(input, length, encoded_buffer2, encoded_buffer_length);
+	if (length == 254 && !memchr(input, 0x00, length)) {
+		/* The normal encoder writes an additional 0x01, if the
+		 * previous code was 0xff and there is no data left.
+		 */
+		zassert_equal(encoded_length2, encoded_length);
+		zassert_equal(encoded_buffer[encoded_length - 1], 0x01);
+		zassert_mem_equal(encoded_buffer2, encoded_buffer, encoded_length - 1);
+	} else {
+		zassert_equal(encoded_length2, encoded_length + 1);
+		zassert_mem_equal(encoded_buffer2, encoded_buffer, encoded_length);
+	}
+
+	size_t decoded_length2;
+	ret = cobs_decode_stream_simple(encoded_buffer2, encoded_length2, decoded_buffer2,
+					length + 1, &decoded_length2);
+	zassert_ok(ret);
+	zassert_equal(decoded_length2, length);
+	zassert_mem_equal(decoded_buffer2, input, length);
+	zassert_equal(decoded_buffer2[length], 0xAB);
+
+	/* Just to double-check, compare to the output of the other
+	 * implementation.
+	 */
+	zassert_equal(decoded_length2, decoded_length);
+	zassert_mem_equal(decoded_buffer2, decoded_buffer, length);
+
 	/* Not run if the tests fail, oh well */
+	free(encoded_buffer2);
+	free(decoded_buffer2);
 	free(encoded_buffer);
 	free(decoded_buffer);
 }
 
 ZTEST(lib_cobs_test, test_single_null)
 {
 	static const uint8_t buffer[] = {0};
-	uint8_t encoded_buffer[3] = {0xAB, 0xAB, 0xAB};
-	static const uint8_t expected_buffer[3] = {1, 1, 0xAB};
+	uint8_t encoded_buffer[] = {0xAB, 0xAB, 0xAB, 0xAB};
+	static const uint8_t expected_buffer[] = {1, 1, 0xAB};
+	static const uint8_t expected_buffer2[] = {1, 1, 0x00, 0xAB};
 
 	const size_t encoded_length = cobs_encode(buffer, sizeof(buffer), encoded_buffer);
-
 	zassert_equal(encoded_length, 2);
-	zassert_mem_equal(encoded_buffer, expected_buffer, sizeof(encoded_buffer));
+	zassert_mem_equal(encoded_buffer, expected_buffer, sizeof(expected_buffer));
+
+	const size_t encoded_length2 = cobs_encode_stream_simple(
```

Comment:
IMO, before running `cobs_encode_stream_simple` we should reset `encoded_buffer` to {0xAB, 0xAB, 0xAB, 0xAB}, because it was already altered by `cobs_encode`

Reply:
same comment applies for `test_hex1`

Reply:
done. I've also added a streaming test to the test to test_255_bytes_null_end since I missed that one.